src/server/pachyderm_test.go

```diff
@@ -6317,6 +6317,87 @@ func TestGetFileWithEmptyCommits(t *testing.T) {
 	require.True(t, strings.Contains(err.Error(), "not found"))
 }
 
+// TestDeleteCommitPropagation deletes an input commit and makes sure all
+// downstream commits are also deleted.
+// DAG in this test: repo -> pipeline[0] -> pipeline[1]
+func TestDeleteCommitPropagation(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping integration tests in short mode")
+	}
+
+	c := getPachClient(t)
+	defer require.NoError(t, c.DeleteAll())
+
+	// Create an input repo
+	repo := uniqueString("TestDeleteCommitPropagation")
+	require.NoError(t, c.CreateRepo(repo))
+
+	pipeline := make([]string, 2)
+	for i := 0; i < 2; i++ {
+		pipeline[i] = uniqueString(fmt.Sprintf("pipeline%d_", i))
+		input := []string{repo, pipeline[0]}[i]
+		require.NoError(t, c.CreatePipeline(
+			pipeline[i],
+			"",
+			[]string{"bash"},
+			[]string{"cp /pfs/*/* /pfs/out/"},
+			&pps.ParallelismSpec{
+				Constant: 1,
+			},
+			client.NewAtomInput(input, "/*"),
+			"",
+			false,
+		))
+	}
+
+	commit := make([]*pfs.Commit, 2)
+	var err error
+	for i := 0; i < 2; i++ {
+		commit[i], err = c.StartCommit(repo, "master")
+		require.NoError(t, err)
+		_, err = c.PutFile(repo, commit[i].ID, "file", strings.NewReader("foo"))
+		require.NoError(t, err)
+		require.NoError(t, c.FinishCommit(repo, commit[i].ID))
+		commitIter, err := c.FlushCommit([]*pfs.Commit{commit[i]}, nil)
+		require.NoError(t, err)
+		commitInfos := collectCommitInfos(t, commitIter)
+		require.Equal(t, 2, len(commitInfos))
+	}
+
+	// Delete the first commit in the input repo (not master, but its parent)
+	// Make sure that 'repo' and all downstream repos only have one commit now
+	commits, err := c.ListCommit(repo, "master", "", 0)
+	require.NoError(t, err)
+	require.Equal(t, 2, len(commits))
+	require.NoError(t, c.DeleteCommit(repo, commit[0].ID))
+	for _, r := range []string{repo, pipeline[0], pipeline[1]} {
+		commits, err := c.ListCommit(r, "master", "", 0)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(commits))
+	}
+
+	// Delete the second commit in the input repo (master)
+	// Make sure that 'repo' and all downstream repos have no commits
+	require.NoError(t, c.DeleteCommit(repo, "master"))
+	for _, r := range []string{repo, pipeline[0], pipeline[1]} {
+		commits, err := c.ListCommit(r, "master", "", 0)
+		require.NoError(t, err)
+		require.Equal(t, 0, len(commits))
+	}
+
+	// Make one more input commit, to be sure that the branches are still
+	// connected properly
+	finalCommit, err := c.StartCommit(repo, "master")
+	require.NoError(t, err)
+	_, err = c.PutFile(repo, finalCommit.ID, "file", strings.NewReader("foo"))
+	require.NoError(t, err)
+	require.NoError(t, c.FinishCommit(repo, finalCommit.ID))
+	commitIter, err := c.FlushCommit([]*pfs.Commit{finalCommit}, nil)
+	require.NoError(t, err)
+	commitInfos := collectCommitInfos(t, commitIter)
+	require.Equal(t, 2, len(commitInfos))
+}
```

Comment:
Could we have a test with the metadata repo case too? I.e. where one commit has several downstream subvenant commits?

Reply:
Just to update: I did write this test and it exposed a bug that I'm currently working on fixing. The bug is:
1. `deleteCommit` calls `createBranch` (to update branches pointing at now-deleted commits)
2. `createBranch` may inspect the HEAD commit of upstream branches while deciding whether to create a new HEAD commit on itself
3. Those upstream branches may not have been updated yet (i.e. may still point at deleted commits) causing `createBranch` to error

I could sort branches by provenance before updating them, but since I actually don't think we want `deleteCommit` to ever cause any pipelines to run, I'm modifying `deleteCommit` so that it doesn't call `createBranch`. I'm still fixing a few bugs in this branch after this change, but it's nearly done.

Reply:
Done

Reply:
I feel like it actually might make sense for `DeleteCommit` to start jobs sometimes. For example suppose I've got a branch with 20 commits on it, I make a pipeline that points at the branch. This should start a job for the head commit, then I delete the head commit. That should cancel the job, but I'd presumably want to then start a job for the new head commit, otherwise I won't be able to see any output from the pipeline until I do another commit.

Reply:
This is now implemented